### ISSUE TITLE
Element is focused before highlight moves, using callbacks correctly.

### ISSUE THREAD

**Issue by rouilj** [kamranahmedse/driver.js#368]
Hello:

I am using driver 1.0.1. I have a form where driver highlights the current form input and provides extra info needed for filling out the form. I have the following two callbacks defined for the driver (they were originally created for 0.X of driver.js and updated for the newer API):
```
       onHighlightStarted: (function (Element) {
            // set focus to highlighted element
            Element.focus();
            // Add listener to catch tab to act like right arrow.
            // Make it easier for users to tab between fields.
            // Add listener when first element is highlighted
            // Add at document level, trying to add at element
            //   seems to not work.
            if ( ! driver.hasPreviousStep() ) { // could be .isFirstStep with 1.x
                document.addEventListener('keyup', checkTabPress, false);
            }
        }),
        onDeselected: (function (Element) {
            // Remove tab listener when driver is no longer activated.
            if ( ! driver.isActive()  ) {
                document.removeEventListener('keyup',
                                             checkTabPress);
            }
        }),
```
with
```
function checkTabPress(e) {
    var ele = document.activeElement;
    console.log("check tab" + e.shiftKey);
    if (e.key === 'Tab') {
        if ( ! e.shiftKey ) {
            console.log("move next");
            driver.moveNext();
        } else {
            console.log("move prev");
            driver.movePrevious();
        }
    }
};
```
This allows movement between form field using tab/shift-tab in addition to keyboard left/right arrow or popup prev/next buttons.

I have focus styles that activate when the input is focused. It looks like the input field gets focus before the
highlight moves. This happens for button, arrow and tab triggers. I thought this was due to using onHighlightStart to set focus rather than using onHighlighted, however using:
```
        onHighlightStarted: (function (Element) {
            // Add listener to catch tab to act like right arrow.
            // Make it easier for users to tab between fields.
            // Add listener when first element is highlighted
            // Add at document level, trying to add at element
            //   seems to not work.
            if ( ! driver.hasPreviousStep() ) {
                document.addEventListener('keyup', checkTabPress, false);
            }
        }),
        onHighlighted: (function (Element) {
            // set focus to highlighted element
            Element.focus();
        }),
```
works correctly for buttons and arrows. The highlight moves and then the focus ring is shown. But when tabbing, I see the focus ring on the next input field before the highlight moves to it. I have tried adding:

```
    if (e.key === 'Tab') {
        e.preventDefault();
        e.stopPropagation();
        if ( ! e.shiftKey ) {
```
to checkTabPress but the focus ring still moves before the highlight.

Any ideas here?

Also should I be using onDestroyStarted when I want to remove the listener? Is that called only when
exiting from the driver (via escape, click on backdrop etc?). Is there an equivalent callback that is called just
before the first item is highlighted? Is this the sequence of callbacks:
```
   driver.drive()
        (should there be an onCreate, onStart, or onDrive callback here?)
        onHighlightStart
        onPopoverRender
        onHighlighted
     movePrevious()/moveNext() via arrow key, prev/next button or called
        onDeselected
        onHighlightStart
        onPopoverRender
        onHighlighted
     destroy() via onCloseClick (or esc key, click on backdrop etc.)
        onDeselected
        onDestroyStarted  - what can/should I do here rather than in onDestroyed?
        onDestroyed
```
Is this right? (Also the API listing seems to be missing the destroy() method).

It looks like there is a missing onCreated/onStart/onDrive callback called when driver first starts. It seems that it would be the compliment to onDestroyed.

Currently I am using: driver.hasPreviousStep() to add the event listener only if I am at the first step of the driver which I assume means the driver just started. However it won't work if I call drive(2) for example to start in the middle of the form.
Also going backwards to the first step with prev adds another copy of the event listener.

Thoughts?

Thanks.

-- rouilj

**Comment by kamranahmedse** [comment]
Instead of registering listeners for `Tab` key, why not simply listen to `focus` and `blur` outside the driver lifecycle and control it simply using `highlight` instead of using tour so you don't have to worry about `moveNext` and `movePrevious`.

I would recommend you to get rid of your old logic and rethink according to the latest version. Most of what you have been doing, you can do it using simple `highlight` and pre-existing hooks. For example here is [a sample implementation for the form](https://driverjs.com/docs/simple-highlight/) similar to what you have been building.

Also, please look at the documentation for the configuration details.

**Comment by rouilj** [comment]
Hello Kamran:

Thanks for your reply.

I chose tour to make sure that the user goes through every form element. Using tab/shift-tab was a response to users expecting to tab to the next field. Because driver can't make the underlying page inert, using tab caused the focus to shift, but tour's highlight/popup did not change.

Your rewrite includes a progress indicator. This is something that was missing before and that I want to keep. From what I can see there is no progress indicator (nor can there be) in the example you recommend.

I read and re-read the documentation of the configuration details and I still don't understand why some of them have Started forms and other's don't. E.G. There is an onHighlightStarted but no onDeselectStarted. Also I don't understand what should be done in a Started versus a non-started callback. The config file documentation:
```
  // Hooks to run before and after highlighting
  // each step. Each hook receives the following
  // parameters:
  //   - element: The target DOM element of the step
  //   - step: The step object configured for the step
  //   - options.config: The current configuration options
  //   - options.state: The current state of the driver
```
doesn't explain what happens between onHighlightStarted and onHighlighted and why I would use one over the other. Similarly for onDestroyStarted and onDestroy.

I also don't understand the exact order of the events/callbacks/hooks. There is no explanation of when onPopoverRender is run compared to the onHighlight* callbacks. So I set up a possible call sequence for you to verify.

As I mentioned, it seems like there is a missing hook that is the opposite of onDestroyed.

I could:

  * define each callback and log them to try to discover the order, or
  * read all the code and try to figure out what the difference is between the onXStarted and onX callbacks.

Even if I were to expend that effort, I have no confidence that I am using them correctly without documentation. Without documentation, I could be relying on a bug in the code that will be changed by some future development.

Thanks again for writing driver.js.
